VERSIONING.md

```diff
@@ -0,0 +1,128 @@
+# Versioning
+
+This document covers Vector's versioning and what it means as a user of Vector.
+
+**Please note, Vector is currently in its pre-1.0 phase and quickly approaching
+1.0. Minor version increments can introduce breaking changes during this phase.
+Please see the [what to expect](#what-to-expect) section for more info.**
+
+<!-- MarkdownTOC autolink="true" style="ordered" indent="   " -->
+
+1. [Convention](#convention)
+1. [Public API](#public-api)
+   1. [Areas that *are* covered](#areas-that-are-covered)
+      1. [Intended for *public* consumption](#intended-for-public-consumption)
+      1. [Intended for *private* consumption](#intended-for-private-consumption)
+   1. [Areas that are *NOT* covered](#areas-that-are-not-covered)
+1. [What to expect](#what-to-expect)
+   1. [Patch and minor versions](#patch-and-minor-versions)
+   1. [Major versions \(breaking changes\)](#major-versions-breaking-changes)
+   1. [Pre 1.0 version](#pre-10-version)
+
+<!-- /MarkdownTOC -->
+
+## Convention
+
+Vector adheres to the [Semantic Versioning 2.0] convention. In summary:
+
+* Versions follow the `MAJOR.MINOR.PATCH` format (i.e., `2.5.1`)
+* `PATCH` increments only when backward compatible bug fixes are introduced.
+* `MINOR` increments only when new, backward compatible functionality is introduced.
+* `MAJOR` increments if any backwards incompatible changes are introduced.
+* Pre `1.0` (major version `0`) is for initial development and `MINOR` version bumps can introduce breaking changes.
+
+## Public API
+
+Semantic Versioning hinges on Vector's defintion of "public API". By nature of
+Vector - a tool that collects, processes, and routes data from disparate systems
+- it has a very large public surface area. It's not immediately obvious which
+parts are covered under our versioning contract and how they're covered. This
+section aims to remove all ambiguity in this area.
+
+### Areas that *are* covered
+
+The following Vector areas are covered in Vector's definition of public API.
+
+#### Intended for *public* consumption
+
+The follow Vector areas are inteded for *public* consumption (consumption by
+anything other than Vector itself). Backward incompatble changes will trigger
+a major version increment.
+
+* [CLI]
```

Comment:
Isn't that assume with just "CLI"?

Reply:
Stealing from the VRL dev guide: explicit > implicit

Reply:
Yeah, agreed, CLI can imply many things. I think it is best to be specific about the fact that it is flags (and maybe exit codes?), but not the log output, for example.

Reply:
In which case I think we can probably drop:

```
  * The root [`vector` command], its input/output.
  * The [`vector validate` subcommand], the exit codes.
```

Below since mentioning that flags and exit codes are covered, but not output, would cover all commands.

Reply:
Fixed in https://github.com/timberio/vector/pull/8795/commits/cc98a20c3d3749121ec0cf6291a5e6715543130b